mlx/io/json.cpp

```diff
@@ -0,0 +1,281 @@
+// Copyright © 2025 Apple Inc.
+
+#include "mlx/io/json.h"
+
+#include <sstream>
+
+namespace mlx::core {
+
+namespace io {
+
+std::string read_digits(std::istream& s) {
+  std::string num = "";
+  char ch = s.get();
+  while (std::isdigit(ch) || ch == '-' || ch == '.' || ch == 'e' || ch == 'E') {
+    num += ch;
+    ch = s.get();
+  }
+  s.seekg(-1, std::ios::cur);
+  return num;
+}
+
+json parse_json_number(std::istream& s) {
+  auto num = read_digits(s);
+  if (num.find_first_of(".eE") != std::string::npos) {
+    return json(std::stod(num));
+  } else {
+    return json(std::stol(num));
+  }
```

Comment:
```suggestion
  size_t pos;
  json ret;
  if (num.find_first_of(".eE") != std::string::npos) {
    ret = json(std::stod(num, &pos));
  } else {
    ret = json(std::stol(num, &pos));
  }
  if (pos != num.size()) {
    throw std::invalid_argument(...);
  }
  return std::move(ret);
```

The above is a bit of a problem because `read_digits` reads any character in `[0-9-eE.]` which means it could be not a number eg `1.2.3.4` and in this case we should fail. `std::stod` and friends return the number of characters processed so it is a fairly easy change to check for correctness.